ls1intum/Artemis
# Quiz Performance and Usability Improvements

# Checklist

### Motivation and Context
We had performance issues with quizzes due to many reloads and a full inbound WS queue due to slow communication in the outbound WS queue. 

### Description
We investigated the issues in more detail and significantly improved the performance of the quiz submission from a few milliseconds (sometimes hundreds of milliseconds) to a few microseconds. So without taking the time of the JSON serialization / deserialization into account, quiz submissions should be 1000x faster than previously 🥇 

### We changed the following:
- [X] We removed the confirmation (outbound) WS message for quiz submissions to lower the stress of the outbound WS queue and make the submissions faster. This means automatic quiz submissions (i.e. when users change their answers in the quiz) are not confirmed any more. In case the client is connected via WS, we assume that saving worked correctly without problems. In case the client is disconnected via WS, automatic saving fails which is displayed. This change might slightly decrease the reliability of the automatic save mechanism.
- [X] To deal better with cases, where the inbound WS queue is not empty, we now use a REST call, when the user clicks on Submit.
- [X] We increased the grace period for quizzes, in case of delayed messages to the server, more quiz submissions will be counted
- [X] We improved the caching of exercises to reduce the database impact in quiz submissions
- [X] We use `saveAll` instead of a for loop and `save` to improve the database write performance when users have submitted their quiz answers
- [X] We added a refresh button for visible quizzes without start date (that have not yet started). In case the quiz starts while the user is disconnected, the user can press the refresh button and does not have to reload the page
- [X] We increased the font size for the Disconnect warning
- [X] When a quiz is visible, but not yet started in the quiz participation page, it can happen that the client is disconnected from the server. In case the client is connected again, it should automatically refresh the page so that students don't miss the quiz in such cases
- [X] Improve error handling when quiz submit does work. Currently, a generic message is shown in the client, this should be improved though

Additional code improvements:
- [X] Fix the REST route for re-evaluate


### Steps for Testing
* Participate in live quizzes in different ways:
1) Start the quiz manually (instructor) while it is open (student) --> it should start automatically
1a) Do the same while being offline (student), go back online, the quiz should start automatically when you have reconnected
2) Start the quiz manually (instructor) while you are in the exercise list of the course (student) --> if you navigate into the quiz, it should start
3) Configure a quiz to start automatically (instructor) --> navigate in the quiz as student (it should start at the start time of the quiz)
3a) Do the same while being offline (student), go back online, the quiz should start automatically when you have reconnected
* Try out different scenarios for disconnects:
1) Disconnect before the quiz starts and reconnect after quiz start
2) Disconnect during the quiz (keep changing answers) and reconnect before the quiz ends (the last version should be saved automatically then)
3) Disconnect during the quiz (keep changing answers) and reconnect after the quiz ends.
==> Check that the messages shown to the user make sense and include useful information.
* Check other edge cases and inspect the network traffic so that no information is processed that the student should not see (e.g. answers to quiz questions before the quiz ends and quiz questions before the quiz starts)
